common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -325,7 +326,7 @@ DefaultHandle<T> claim() {
             if (batch.isEmpty()) {
                 handles.drain(this, chunkSize);
             }
-            DefaultHandle<T> handle = batch.pollFirst();
+            DefaultHandle<T> handle = batch.isEmpty() ? null : batch.remove(0);
```

Comment:
To use is a queue, it should have been batch.remove(batch.size() - 1)

Said that, maybe just using ArrayDeque::pollLast could be enough, looking at the implementation in the jdk repository